### ISSUE TITLE
Cannot use programmatically created/modified DatabaseChangeLog in Liquibase.update()

### ISSUE THREAD

**Issue by tonketonky** [liquibase/liquibase#4657]
### Search first

- [X] I searched and no similar issues were found

### Description

Hi team,

we tried upgrading Liquibase library from `4.20.0` to `4.23.0` and found out that since version `4.21.0` it is not possible to use programmatically created/modified `DatabaseChangeLog` instance when executing `Liquibase.update()`.

When using `liquibase:4.20.0` it is possible to do following:
```
try (Connection connection = DriverManager.getConnection(databaseUrl, username, password)) {
    Database database = DatabaseFactory.getInstance().findCorrectDatabaseImplementation(new JdbcConnection(connection));

    DatabaseChangeLog databaseChangeLog = new DatabaseChangeLog();

    // build up databaseChangeLog instance here

    Liquibase liquibase = new Liquibase(databaseChangeLog, new ClassLoaderResourceAccessor(), database);
    liquibase.update(); // execute with programmatically created DatabaseChangeLog
}
```

and also possible to do following:
```
try (Connection connection = DriverManager.getConnection(databaseUrl, username, password);
    Liquibase liquibase = new Liquibase(changeLogFilePath, new ClassLoaderResourceAccessor(), connection)) {

    DatabaseChangeLog databaseChangeLog = liquibase.getDatabaseChangeLog();

    // modify databaseChangeLog instance here

    liquibase.update(); // execute with programmatically modified DatabaseChangeLog 
}
```

When using `liquibase:4.21.0` (and higher) above examples does not work.

I dove into the code and can see the root cause. 

In `liquibase:4.20.0` in `Liquibase` class there is this point how `databaseChangeLog` for further use during update process is retrieved: 
```
    public void update(Contexts contexts, LabelExpression labelExpression, boolean checkLiquibaseTables) throws LiquibaseException {
        runInScope(() -> {
            ...

            Operation updateOperation = null;
            BufferedLogService bufferLog = new BufferedLogService();
            DatabaseChangeLog changeLog;
            HubUpdater hubUpdater = null;
            try {
                changeLog = getDatabaseChangeLog();  // <--- THIS
                ...
```
Method `getDatabaseChangeLog()` returns `this.databaseChangeLog` if it is already initialized - this is the programmatically created/modified `databaseChangeLog` from above examples. Everything works as expected.
```
private DatabaseChangeLog getDatabaseChangeLog(boolean shouldWarnOnMismatchedXsdVersion) throws LiquibaseException {
    if (databaseChangeLog == null && changeLogFile != null) {
        Scope.getCurrentScope().addMdcValue(MdcKey.CHANGELOG_FILE, changeLogFile);
        ChangeLogParser parser = ChangeLogParserFactory.getInstance().getParser(changeLogFile, resourceAccessor);
        if (parser instanceof XMLChangeLogSAXParser) {
            ((XMLChangeLogSAXParser) parser).setShouldWarnOnMismatchedXsdVersion(shouldWarnOnMismatchedXsdVersion);
        }
        databaseChangeLog = parser.parse(changeLogFile, changeLogParameters, resourceAccessor);
    }

    return databaseChangeLog;
}
```

However, In `liquibase:4.21.0` in `Liquibase` class this is how update process is being started:
```
public void update(Contexts contexts, LabelExpression labelExpression, Writer output, boolean checkLiquibaseTables)
        throws LiquibaseException {
    runInScope(() -> {
        CommandScope updateCommand = new CommandScope(UpdateSqlCommandStep.COMMAND_NAME);
        updateCommand.addArgumentValue(DbUrlConnectionCommandStep.DATABASE_ARG, getDatabase());
        updateCommand.addArgumentValue(UpdateSqlCommandStep.CHANGELOG_FILE_ARG, changeLogFile);
        updateCommand.addArgumentValue(UpdateSqlCommandStep.CONTEXTS_ARG, contexts != null ? contexts.toString() : null);
        updateCommand.addArgumentValue(UpdateSqlCommandStep.LABEL_FILTER_ARG, labelExpression != null ? labelExpression.getOriginalString() : null);
        updateCommand.addArgumentValue(ChangeExecListenerCommandStep.CHANGE_EXEC_LISTENER_ARG, changeExecListener);
        updateCommand.addArgumentValue(DatabaseChangelogCommandStep.CHANGELOG_PARAMETERS, changeLogParameters);
        updateCommand.setOutput(new WriterOutputStream(output, GlobalConfiguration.OUTPUT_FILE_ENCODING.getCurrentValue()));
        updateCommand.execute();
    });
}
``` 
We can see that only `this.changeLogFile` (i.e. path change log file) is being passed and actual `this.databaseChangeLog` which may be created/modified programmatically is not used at all.

Then later there is this point in `DatabaseChangeLogCommandStep.run()` method where `DatabaseChangelogCommandStep.getDatabaseChangeLog()` is invoked:
```
public void run(CommandResultsBuilder resultsBuilder) throws Exception {
    ...
    changeLogParameters.setLabels(labels);
    addCommandFiltersMdc(labels, contexts);

    DatabaseChangeLog databaseChangeLog = getDatabaseChangeLog(changeLogFile, changeLogParameters);   // <--- THIS
    ...
```

However, `DatabaseChangelogCommandStep.getDatabaseChangeLog()` method parses `databaseChangeLog` from the file at provided path. This has 2 implications:
- when programmatically creating `databaseChangeLog` and not providing change log file path, update process will not use provided `databaseChangeLog` but rather it will try to parse change log from file and will fail
- when programmatically modifying `databaseChangeLog`. update process will just parse change log from the file and all modifications to `databaseChangeLog` are lost.

### Steps To Reproduce

- programmatically creating `DatabaseChangeLog`:
```
try (Connection connection = DriverManager.getConnection(databaseUrl, username, password)) {
    Database database = DatabaseFactory.getInstance().findCorrectDatabaseImplementation(new JdbcConnection(connection));

    // 1. create DatabaseChangeLog instance
    DatabaseChangeLog databaseChangeLog = new DatabaseChangeLog();

    // 2. build up databaseChangeLog instance here

    Liquibase liquibase = new Liquibase(databaseChangeLog, new ClassLoaderResourceAccessor(), database);

    // 3. execute with programmatically created DatabaseChangeLog
    liquibase.update();
}
```
- programmatically modifying `DatabaseChangeLog`:
```
try (Connection connection = DriverManager.getConnection(databaseUrl, username, password);

    // 1. create Liquibase instance and provide change log file path
    Liquibase liquibase = new Liquibase(changeLogFilePath, new ClassLoaderResourceAccessor(), connection)) {

    DatabaseChangeLog databaseChangeLog = liquibase.getDatabaseChangeLog();

    // 2. modify databaseChangeLog instance here

    // 3. execute with programmatically modified DatabaseChangeLog 
    liquibase.update(); 
}
```

### Expected/Desired Behavior

Update process uses programmatically created/modified `DatabaseChangeLog` instance

### Liquibase Version

4.21.0

### Database Vendor & Version

PostgreSQL 11.17

### Liquibase Integration

manual execution

### Liquibase Extensions

liquibase-core

### OS and/or Infrastructure Type/Provider

Ubuntu 20.04.6 LTS

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

**Comment by tati-qalified** [comment]
Hello @tonketonky, thank you for bringing this to our attention.

I need some info in order to test this issue: 
- are you using any frameworks (spring boot, hibernate, JPA, etc.)?
- are you using any build automation tools (maven, gradle, etc.)?
- are there any other details about the project's configuration or pipeline, outside of the code you've provided, that you think could be involved in the issue or would be relevant?

Thanks in advance,
Tatiana

**Comment by tonketonky** [comment]
Hello @tati-qalified, I apology for the late reply. Here are answers for your questions:
- we are using spring boot and hibernate
- we are using maven
- I don't think there's more details to it. In principle, we are triggering `liquibase.update()` manually, not from an automated pipeline and we are using liquibase programmtically as I described above.

Thanks in advance for any investigation of this,
Tony

**Comment by tonketonky** [comment]
We just upgraded to `liquibase:4.26.0` and found out this issue has been fixed since `liquibase:4.24.0`. Closing issue.
